### PR TITLE
Updated circular_linked_list

### PR DESCRIPTION
**Pull request by shellhub**
### **Describe your change:**

* [x] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [x] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: data_structures/linked_list/circular_linked_list.py [hunk 1]
```diff
@@ -2,201 +2,137 @@
...
-        >>> print(f"{len(cll)}: {cll}")
-        1: <Node data=1>
-        >>> cll.append(2)
-        >>> print(f"{len(cll)}: {cll}")
-        2: <Node data=1> => <Node data=2>
-        """
-        current_node = self.head
+    def __iter__(self):
+        node = self.head
+        while self.head is not None:
```
Comment (cclauss) [inline data_structures/linked_list/circular_linked_list.py:17]:
```suggestion
        while self.head:
```

File: data_structures/linked_list/circular_linked_list.py [hunk 2]
```diff
@@ -2,201 +2,137 @@
...
-        >>> cll.prepend(2)
-        >>> print(f"{len(cll)}: {cll}")
-        2: <Node data=2> => <Node data=1>
-        """
-        current_node = self.head
+    def insert_head(self, data: Any) -> None:
+        self.insert_nth(0, data)
 
+    def insert_nth(self, index: int, data: Any) -> None:
+        if index < 0 or index > len(self):
```
Comment (cclauss) [inline data_structures/linked_list/circular_linked_list.py:36]:
```suggestion
        if not 0 <=index <= len(self):
```

**Review by cclauss — CHANGES_REQUESTED** [review]
It feels like we removed a lot of tests and did not add new ones.

**Comment by cclauss** [comment]
What parts of this code have no tests?

**Comment by shellhub** [comment]
I test all fuctions in test_circular_linked_list() function.

**Comment by shellhub** [comment]
I don't test `print_list()` function.  We do need test it ? @cclauss

File: data_structures/linked_list/circular_linked_list.py [hunk 3]
```diff
@@ -2,201 +2,137 @@
...
+
+    try:
+        circular_linked_list.delete_nth(0)
+        assert False
+    except IndexError:
+        assert True
+
+    for i in range(5):
+        assert len(circular_linked_list) == i
+        circular_linked_list.insert_nth(i, i + 1)
```
Comment (cclauss) [inline data_structures/linked_list/circular_linked_list.py:121]: Please add another is_empty() test.

**Comment by cclauss** [comment]
Please delete `.print_list()` as it is completely redundant.

File: data_structures/linked_list/circular_linked_list.py (hunk 3 above)
Comment (shellhub) [reply data_structures/linked_list/circular_linked_list.py:121]: Like this ?
``` python
    assert circular_linked_list.is_empty() is True
    for i in range(5):
        assert len(circular_linked_list) == i
        circular_linked_list.insert_nth(i, i + 1)
```

File: data_structures/linked_list/circular_linked_list.py [hunk 4]
```diff
@@ -2,201 +2,137 @@
...
+                temp = temp.next
+            delete_node = temp.next
+            temp.next = temp.next.next
+            if index == len(self) - 1:  # delete at tail
+                self.tail = temp
+        return delete_node.data
 
-            temp_node.next_ptr = current_node.next_ptr
+    def print_list(self) -> None:
+        print(self)
```
Comment (shellhub) [inline data_structures/linked_list/circular_linked_list.py:81]:
```suggestion
```

File: data_structures/linked_list/circular_linked_list.py [hunk 5]
```diff
@@ -2,201 +2,137 @@
...
+    except IndexError:
+        assert True
+
+    try:
+        circular_linked_list.delete_nth(0)
+        assert False
+    except IndexError:
+        assert True
+
+    for i in range(5):
```
Comment (shellhub) [inline data_structures/linked_list/circular_linked_list.py:119]:
```suggestion
    assert circular_linked_list.is_empty() is True
    for i in range(5):
```

**Review by shellhub — COMMENTED** [review]
deleted `print_list()` and test `is_empty()` again.

File: data_structures/linked_list/circular_linked_list.py (hunk 5 above)
Comment (cclauss) [reply data_structures/linked_list/circular_linked_list.py:117]: No.  We are looking for a test where .is_empty() returns False.  We already have one that returns True.

File: data_structures/linked_list/circular_linked_list.py [hunk 6]
```diff
@@ -2,201 +2,135 @@
...
+    circular_linked_list.insert_head(0)
+    assert str(circular_linked_list) == "->".join(str(i) for i in range(0, 7))
+
+    assert circular_linked_list.delete_front() == 0
+    assert circular_linked_list.delete_tail() == 6
+    assert str(circular_linked_list) == "->".join(str(i) for i in range(1, 6))
+    assert circular_linked_list.delete_nth(2) == 3
+
+    circular_linked_list.insert_nth(2, 3)
+    assert str(circular_linked_list) == "->".join(str(i) for i in range(1, 6))
```
Comment (shellhub) [inline data_structures/linked_list/circular_linked_list.py:134]:
```suggestion

assert circular_linked_list.is_empty() is False

```

File: data_structures/linked_list/circular_linked_list.py [hunk 7]
```diff
@@ -2,201 +2,135 @@
...
+    except IndexError:
+        assert True
+
+    try:
+        circular_linked_list.delete_nth(0)
+        assert False
+    except IndexError:
+        assert True
+
+    assert circular_linked_list.is_empty() is True
```
Comment (shellhub) [inline data_structures/linked_list/circular_linked_list.py:116]:
```suggestion
```

**Comment by shellhub** [comment]
Done. @cclauss

File: data_structures/linked_list/circular_linked_list.py [hunk 8]
```diff
@@ -2,201 +2,138 @@
...
+    assert circular_linked_list.delete_front() == 0
+    assert circular_linked_list.delete_tail() == 6
+    assert str(circular_linked_list) == "->".join(str(i) for i in range(1, 6))
+    assert circular_linked_list.delete_nth(2) == 3
+
+    circular_linked_list.insert_nth(2, 3)
+    assert str(circular_linked_list) == "->".join(str(i) for i in range(1, 6))
+
+
+assert circular_linked_list.is_empty() is False
```
Comment (shellhub) [inline data_structures/linked_list/circular_linked_list.py:136]:
```suggestion
    assert circular_linked_list.is_empty() is False
```

**Review by shellhub — COMMENTED** [review]
fixed indentation

**Comment by shellhub** [comment]
@cclauss Please review again :)

**Review by cclauss — APPROVED** [review]
Nice work!

**Pull request merged**